### ISSUE TITLE
S3 buckets with dot in name

### ISSUE THREAD

**Issue by andrew-slater** [samtools/htslib#764]
I'm able to run htsfile successfully on an s3 bucket without a dot in the name:
```
$ ./htsfile -h s3://omicsofttest1/10samp.10kvar.vcf.gz
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
```
But when the bucket has a dot in the name, it does not work:
```
$ ./htsfile -h s3://east.test.omicsoft/Andrew/10000samp.1mvar.vcf.gz
htsfile: can't open "s3://east.test.omicsoft/Andrew/10000samp.1mvar.vcf.gz": Input/output error
```
With more verbose, seems the problem is with the bucket name in the domain:
```
$ ./htsfile -h -v -v -v -v -v -v s3://east.test.omicsoft/Andrew/10000samp.1mvar.vcf.gz
[D::init_add_plugin] Loaded "knetfile"
[D::init_add_plugin] Loaded "mem"
[D::init_add_plugin] Loaded "/home/andrews/Temp/hgslib-1.9/install_dir/libexec/htslib/hfile_libcurl.so"
[M::hfile_gcs.init] version 1.9
[D::init_add_plugin] Loaded "/home/andrews/Temp/hgslib-1.9/install_dir/libexec/htslib/hfile_gcs.so"
[M::hfile_s3.init] version 1.9
[D::init_add_plugin] Loaded "/home/andrews/Temp/hgslib-1.9/install_dir/libexec/htslib/hfile_s3.so"
* About to connect() to east.test.omicsoft.s3.amazonaws.com port 443 (#0)
*   Trying 54.231.49.218...
* Connected to east.test.omicsoft.s3.amazonaws.com (54.231.49.218) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* Server certificate:
*       subject: CN=*.s3.amazonaws.com,O=Amazon.com Inc.,L=Seattle,ST=Washington,C=US
*       start date: Sep 22 00:00:00 2017 GMT
*       expire date: Jan 03 12:00:00 2019 GMT
*       common name: *.s3.amazonaws.com
*       issuer: CN=DigiCert Baltimore CA-2 G2,OU=www.digicert.com,O=DigiCert Inc,C=US
* NSS error -12276 (SSL_ERROR_BAD_CERT_DOMAIN)
* Unable to communicate securely with peer: requested domain name does not match the server's certificate.
* Closing connection 0
htsfile: can't open "s3://east.test.omicsoft/Andrew/10000samp.1mvar.vcf.gz": Input/output error
```
Seems the wildcard domain in the certificate can’t handle dots in the bucket name: https://aws.amazon.com/premiumsupport/knowledge-center/acm-certificate-error-https/
```
If you use a wildcard name (*), the wildcard matches only one subdomain level. For example, *.example.com can protect login.example.com and test.example.com, but the wildcard cannot protect test.login.example.com nor example.com. 
```

**Comment by whitwham** [comment]
These are notes for me to refer to later.  There are two styles of specifying a resource in S3, virtual-style and path-style.

virtual-style:
mybucket.s3.amazonaws.com/myfile

path-style:
s3.amazonaws.com/mybucket/myfile

As stated [here](https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html):
> When using virtual hosted–style buckets with SSL, the SSL wild card certificate only matches buckets that do not contain periods.

If the path-style bucket is used then there is no SSL problem.

Currently htsfile s3 handling code will use virtual-style if it can.  There is code in there to do path-style and we could either make it used by default or used when there is a dot in the bucket name.  However there is a possible problem when using Amazon S3.

> The path-style syntax, however, requires that you use the region-specific endpoint when attempting to access a bucket. 

The htslib s3 handling code uses v2 signatures and does very little (if anything) with region settings.  Updating to v4 signatures would make use of region settings (as well as allowing access to regions that do not use v2 sigs) though it would take some work to get right.

**Comment by daviesrob** [comment]
Fixed by #839
